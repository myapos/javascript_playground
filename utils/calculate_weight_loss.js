/**
 * It will calculate all the data for plotting graphs
 *
 * Parameters:
 * rawData: an array containing objects from csv reading
 * dates: an array containing only the dates from csv reading
 */

import { plot } from 'nodeplotlib';
import addMean from './addMean';
import dateDiffInDays from './dateDiffInDays';
import missingValues from './missingValues';
import verboseLog from './verboseLog';

const calculate_weight_loss = ({ rawData, dates, currentDate }) => {
  let values = [];

  let filledDates = [];
  let filledValues = [];

  // detect any missing dates and fill them
  values = missingValues({ rawData, dates, mode: 'forward', currentDate });
  filledDates = values.map((value) => value.Date);
  filledValues = values.map((value) => {
    const replaced = value['Kgrs'].replace(',', '.');
    return parseFloat(replaced);
  });

  verboseLog(`filledValues ${filledValues}`);
  verboseLog(`filledValues length ${filledValues.length}`);

  verboseLog(`filledDates ${filledDates}`);
  verboseLog(`filledDates length ${filledDates.length}`);

  const midsY = [];
  const midsX = [];

  const slices = [];
  let slice = [];
  let p1 = 0,
    p2 = 0;
  let dif = 7;

  // const lastDateIndexPeriod = Math.floor(filledValues.length / (dif - 1)) * (dif - 1);
  const lastDateIndexPeriod = Math.floor(filledValues.length / dif) * dif;
  verboseLog(`lastDateIndexPeriod: ${lastDateIndexPeriod}`);

  let remainingValues = [];
  filledValues.forEach((value, index) => {
    const difference = dateDiffInDays(filledDates[p1], filledDates[p2]);
    const range = filledDates[p1] + ' to ' + filledDates[p2 - 1];

    // p2 - p1 > dif - 1
    if (difference > dif - 1) {
      slice = filledValues.slice(p1, p2);
      addMean({ slice, midsY, range: range, midsX });
      slices.push(slice);
      p1 = p2;
    }

    if (index === filledValues.length - 1) {
      verboseLog('Calculate last period values');

      // 1. get last midX
      const lastMidsX = midsX[midsX.length - 1].split(' ');
      const dateOfLastPeriod = lastMidsX[lastMidsX.length - 1];
      verboseLog('last period: ' + dateOfLastPeriod);
      let lastDateIndexPeriod;

      // 2. find index of last date period from array
      filledDates.some((date, index) => {
        if (date === dateOfLastPeriod) {
          lastDateIndexPeriod = index + 1;
          return true;
        }
        return false;
      });

      verboseLog(
        'lastDateIndexPeriod ' +
          lastDateIndexPeriod +
          ' last value before ' +
          filledValues[lastDateIndexPeriod],
      );

      // 3. get remaining values  as as slice from filled values and lastDateIndexPeriod
      remainingValues = filledValues.slice(lastDateIndexPeriod, filledValues.length);
      verboseLog(`last iteration remaining values: ${remainingValues}`);

      // 4. push remainingValues in the last iteration
      slices.push(remainingValues);
      addMean({
        slice: remainingValues,
        midsY,
        range: filledDates[lastDateIndexPeriod] + ' to Today',
        midsX,
      });
    }

    // increase counter
    p2++;
  });

  // verboseLog(`remainingValues completed: ${remainingValues}`);
  // verboseLog(`slices: ${slices}`);
  verboseLog(`midsY: ${midsY}`);
  verboseLog(`midsX: ${midsX}`);

  var option2 = process.argv.slice(2)[1];

  if (option2 === 'graphs') {
    /**
     * examples:
     * npm start - graphs
     * npm start '' graphs
     * npm start verbose graphs
     */
    const data = [{ x: midsX, y: midsY, type: 'line' }];
    plot(data);

    // plot initial data
    const data_ = [{ x: filledDates, y: filledValues, type: 'line' }];
    plot(data_);
  }

  return {
    midsX,
    midsY,
    filledDates,
    filledValues,
  };
};

export default calculate_weight_loss;
