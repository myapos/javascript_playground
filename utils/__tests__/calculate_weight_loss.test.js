import calculate_weight_loss from '../calculate_weight_loss';
import { rawData } from './__data__/rawData';
import { dates } from './__data__/dates';
import { bigDataSet } from './__data__/bigDataSet';
import { bigDateSet } from './__data__/bigDateSet';
describe('calculate weight loss graph values', () => {
  it('to should calculate midsX, midsY of graphs and filledValues, filledDates', () => {
    const { midsX, midsY, filledDates, filledValues } = calculate_weight_loss(rawData, dates);

    expect(midsX).toMatchSnapshot();
    expect(midsY).toMatchSnapshot();
    expect(filledDates).toMatchSnapshot();
    expect(filledValues).toMatchSnapshot();
  });

  it('to should calculate midsX, midsY of graphs and filledValues, filledDates with big dataSets', () => {
    const { midsX, midsY, filledDates, filledValues } = calculate_weight_loss(
      bigDataSet,
      bigDateSet,
    );

    expect(midsX).toMatchSnapshot();
    expect(midsY).toMatchSnapshot();
    expect(filledDates).toMatchSnapshot();
    expect(filledValues).toMatchSnapshot();
  });
});
