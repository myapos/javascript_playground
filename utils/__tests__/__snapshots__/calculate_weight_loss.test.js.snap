// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates 1`] = `
Array [
  "14-03-2020 to 20-03-2020",
  "21-03-2020 to 27-03-2020",
  "28-03-2020 to 31-03-2020",
  "03-05-2020 to 09-05-2020",
  "10-05-2020 to 16-05-2020",
  "17-05-2020 to 23-05-2020",
  "24-05-2020 to 30-05-2020",
  "31-05-2020 to Today",
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates 2`] = `
Array [
  108,
  108,
  108,
  104.89999999999999,
  103.97142857142858,
  104.04285714285716,
  103.07142857142858,
  103.1,
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates 3`] = `
Array [
  "14-03-2020",
  "15-03-2020",
  "16-03-2020",
  "17-03-2020",
  "18-03-2020",
  "19-03-2020",
  "20-03-2020",
  "21-03-2020",
  "22-03-2020",
  "23-03-2020",
  "24-03-2020",
  "25-03-2020",
  "26-03-2020",
  "27-03-2020",
  "28-03-2020",
  "29-03-2020",
  "30-03-2020",
  "31-03-2020",
  "03-05-2020",
  "04-05-2020",
  "05-05-2020",
  "06-05-2020",
  "07-05-2020",
  "08-05-2020",
  "09-05-2020",
  "10-05-2020",
  "11-05-2020",
  "12-05-2020",
  "13-05-2020",
  "14-05-2020",
  "15-05-2020",
  "16-05-2020",
  "17-05-2020",
  "18-05-2020",
  "19-05-2020",
  "20-05-2020",
  "21-05-2020",
  "22-05-2020",
  "23-05-2020",
  "24-05-2020",
  "25-05-2020",
  "26-05-2020",
  "27-05-2020",
  "28-05-2020",
  "29-05-2020",
  "30-05-2020",
  "31-05-2020",
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates 4`] = `
Array [
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  103.6,
  103.6,
  103.6,
  103.6,
  103.6,
  103.6,
  104.7,
  104.2,
  104.2,
  104.2,
  103.7,
  103.7,
  103.2,
  102.8,
  103.1,
  103.1,
  103.1,
  103.1,
  103.1,
  103.1,
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates with big dataSets 1`] = `
Array [
  "14-03-2020 to 20-03-2020",
  "21-03-2020 to 27-03-2020",
  "28-03-2020 to 31-03-2020",
  "03-05-2020 to 09-05-2020",
  "10-05-2020 to 16-05-2020",
  "17-05-2020 to 23-05-2020",
  "24-05-2020 to 30-05-2020",
  "31-05-2020 to 06-06-2020",
  "07-06-2020 to 13-06-2020",
  "14-06-2020 to 20-06-2020",
  "21-06-2020 to 27-06-2020",
  "28-06-2020 to 04-07-2020",
  "05-07-2020 to 11-07-2020",
  "12-07-2020 to 18-07-2020",
  "19-07-2020 to 25-07-2020",
  "26-07-2020 to 01-08-2020",
  "02-08-2020 to 08-08-2020",
  "09-08-2020 to 15-08-2020",
  "16-08-2020 to 22-08-2020",
  "23-08-2020 to 29-08-2020",
  "30-08-2020 to 05-09-2020",
  "06-09-2020 to Today",
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates with big dataSets 2`] = `
Array [
  108,
  108,
  108,
  104.89999999999999,
  103.97142857142858,
  104.04285714285716,
  102.8,
  102.65714285714286,
  101.78571428571429,
  102.57142857142856,
  101.2857142857143,
  100.61428571428571,
  99.54285714285713,
  99.92857142857144,
  99.74285714285713,
  99.77142857142859,
  98.85714285714286,
  97.9857142857143,
  97.8142857142857,
  97.92857142857143,
  97.87142857142858,
  96.91428571428573,
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates with big dataSets 3`] = `
Array [
  "14-03-2020",
  "15-03-2020",
  "16-03-2020",
  "17-03-2020",
  "18-03-2020",
  "19-03-2020",
  "20-03-2020",
  "21-03-2020",
  "22-03-2020",
  "23-03-2020",
  "24-03-2020",
  "25-03-2020",
  "26-03-2020",
  "27-03-2020",
  "28-03-2020",
  "29-03-2020",
  "30-03-2020",
  "31-03-2020",
  "03-05-2020",
  "04-05-2020",
  "05-05-2020",
  "06-05-2020",
  "07-05-2020",
  "08-05-2020",
  "09-05-2020",
  "10-05-2020",
  "11-05-2020",
  "12-05-2020",
  "13-05-2020",
  "14-05-2020",
  "15-05-2020",
  "16-05-2020",
  "17-05-2020",
  "18-05-2020",
  "19-05-2020",
  "20-05-2020",
  "21-05-2020",
  "22-05-2020",
  "23-05-2020",
  "24-05-2020",
  "25-05-2020",
  "26-05-2020",
  "27-05-2020",
  "28-05-2020",
  "29-05-2020",
  "30-05-2020",
  "31-05-2020",
  "01-06-2020",
  "02-06-2020",
  "03-06-2020",
  "04-06-2020",
  "05-06-2020",
  "06-06-2020",
  "07-06-2020",
  "08-06-2020",
  "09-06-2020",
  "10-06-2020",
  "11-06-2020",
  "12-06-2020",
  "13-06-2020",
  "14-06-2020",
  "15-06-2020",
  "16-06-2020",
  "17-06-2020",
  "18-06-2020",
  "19-06-2020",
  "20-06-2020",
  "21-06-2020",
  "22-06-2020",
  "23-06-2020",
  "24-06-2020",
  "25-06-2020",
  "26-06-2020",
  "27-06-2020",
  "28-06-2020",
  "29-06-2020",
  "30-06-2020",
  "01-07-2020",
  "02-07-2020",
  "03-07-2020",
  "04-07-2020",
  "05-07-2020",
  "06-07-2020",
  "07-07-2020",
  "08-07-2020",
  "09-07-2020",
  "10-07-2020",
  "11-07-2020",
  "12-07-2020",
  "13-07-2020",
  "14-07-2020",
  "15-07-2020",
  "16-07-2020",
  "17-07-2020",
  "18-07-2020",
  "19-07-2020",
  "20-07-2020",
  "21-07-2020",
  "22-07-2020",
  "23-07-2020",
  "24-07-2020",
  "25-07-2020",
  "26-07-2020",
  "27-07-2020",
  "28-07-2020",
  "29-07-2020",
  "30-07-2020",
  "31-07-2020",
  "01-08-2020",
  "02-08-2020",
  "03-08-2020",
  "04-08-2020",
  "05-08-2020",
  "06-08-2020",
  "07-08-2020",
  "08-08-2020",
  "09-08-2020",
  "10-08-2020",
  "11-08-2020",
  "12-08-2020",
  "13-08-2020",
  "14-08-2020",
  "15-08-2020",
  "16-08-2020",
  "17-08-2020",
  "18-08-2020",
  "19-08-2020",
  "20-08-2020",
  "21-08-2020",
  "22-08-2020",
  "23-08-2020",
  "24-08-2020",
  "25-08-2020",
  "26-08-2020",
  "27-08-2020",
  "28-08-2020",
  "29-08-2020",
  "30-08-2020",
  "31-08-2020",
  "01-09-2020",
  "02-09-2020",
  "03-09-2020",
  "04-09-2020",
  "05-09-2020",
  "06-09-2020",
  "07-09-2020",
  "08-09-2020",
  "09-09-2020",
  "10-09-2020",
  "11-09-2020",
  "12-09-2020",
]
`;

exports[`calculate weight loss graph values to should calculate midsX, midsY of graphs and filledValues, filledDates with big dataSets 4`] = `
Array [
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  108,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  104.9,
  103.6,
  103.6,
  103.6,
  103.6,
  103.6,
  103.6,
  104.7,
  104.2,
  104.2,
  104.2,
  103.7,
  103.7,
  103.2,
  102.8,
  103.1,
  102.8,
  102.1,
  103.1,
  102.5,
  102.5,
  102.9,
  103.5,
  102.6,
  102.6,
  102.6,
  101.9,
  101.9,
  102.7,
  102.7,
  101.6,
  101.6,
  101,
  101,
  103,
  103.2,
  103.2,
  102.9,
  102.4,
  101.3,
  102,
  101.3,
  101.8,
  101.5,
  101.5,
  101.1,
  101.1,
  100.7,
  100.7,
  100.7,
  100.7,
  100.3,
  100.3,
  100.8,
  100.8,
  100.2,
  99.8,
  98.8,
  98.4,
  99.4,
  99.4,
  100.8,
  101.4,
  99.9,
  99.9,
  99.2,
  99.7,
  99.7,
  99.7,
  99.7,
  98.7,
  99.2,
  99.2,
  99.7,
  100.3,
  101.4,
  102.4,
  100.4,
  100.5,
  98.4,
  98.7,
  99,
  99,
  100.3,
  99.2,
  98.5,
  98.5,
  98.5,
  98.5,
  98.5,
  98.5,
  98.5,
  97.9,
  97.3,
  97.5,
  98,
  98.2,
  97.6,
  98,
  97.5,
  97.5,
  98.5,
  97.8,
  97.8,
  97.3,
  98.5,
  98.2,
  97.6,
  97.6,
  98.4,
  97.9,
  97.5,
  98.2,
  98.2,
  98.2,
  97.6,
  97.2,
  98.2,
  97.1,
  97.1,
  96.3,
  96.6,
  97.1,
  97.1,
  97.1,
]
`;
